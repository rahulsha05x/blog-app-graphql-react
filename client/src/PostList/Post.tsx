
export interface Post {
    id:string,
    title:string,
    description:string
}

export class Post implements Post {
    constructor(public id:string,public title:string,public description:string) {}
}
