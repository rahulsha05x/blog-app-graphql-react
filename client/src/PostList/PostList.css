.PostList__Item {
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;
    min-height:200px;
    height:400px;

}
.btn--primary a{
    color:#fff;
}