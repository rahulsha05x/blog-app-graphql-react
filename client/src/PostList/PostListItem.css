.PostItem {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    border:1px solid #dedede;
    border-radius: 2px;
    box-shadow: 5px 5px #dedede;
    align-items: flex-start;
    padding: 10px;
    justify-content: space-between;
    margin:20px;
    min-height: 100px;
    width:550px;
    height:250px;
    align-items: flex-start;
    box-sizing: border-box;
}
.PostItem__Card--Heading {
    width: 500px;
    min-height: 85px;
    font-weight: 600;
}
.PostItem__Card--Description {
    max-width:500px;
    min-height:80px;
    font-weight:400;
}
.PostItem__Card--Heading a{
    color:black;
}
.PostItem__Card__Button {
    display: flex;
    flex-wrap: nowrap;
    margin:10px;
}
.PostItem__Card__Button--edit,.PostItem__Card__Button--delete {
    margin-right:30px;
}