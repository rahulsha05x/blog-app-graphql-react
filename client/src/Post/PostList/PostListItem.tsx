import React from "react";
import './PostListItem.css';
import {  useHistory } from "react-router-dom"
import { Post } from "./Post";
interface Props {
    key:string,
    post:Post,
    edit: (id:string) => void,
    deletePost: (id:string) => void,
}
/**
 * Receives posts array, deletePost and edit.
 * displays title and description for single post
 * contains buttons to edit and delete post
 */
const PostListItem:React.FC<Props> = ({post,deletePost,edit}: Props) => {
  const { id, title, description } = post;
  const history = useHistory()
  const trimDescription = description.length > 100? `${description.substring(0,150)}...`:description;
  /**
 * Navigates to posts/:id to display post
 */
  const openPost = (id:any) => {
    history.push(`/posts/${id}`)
  }
  return (
    <article className='PostItem' data-testid='PostList-Item'>
        <section className='PostItem__Card'>
            <h2 className='PostItem__Card--Heading' title={title}>
                <button type='button' data-testid='postitem-heading' className='PostItem__Card--heading' onClick={()=>openPost(id)}>{title}</button>
            </h2>
            <p className='PostItem__Card--Description'>
                {trimDescription}
            </p>
            <article className='PostItem__Card__Button'>
                <button type='button' 
                    className='PostItem__Card__Button--edit btn--primary'
                    onClick={()=>edit(id)}> Edit</button>
                <button 
                    className='PostItem__Card__Button--delete btn--primary' 
                    type='button'
                    onClick={()=>deletePost(id)}>Delete</button>
            </article>
        </section>

    </article>
  );
};

export default PostListItem;
