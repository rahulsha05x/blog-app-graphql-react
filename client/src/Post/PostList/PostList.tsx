import React from 'react';
import './PostList.css';
import PostListItem from './PostListItem';
import { Row, Col } from 'reactstrap';
import { useHistory } from 'react-router-dom';
import { Post } from './Post';
import { POST_LIST_HEADING, ADD_POST_BUTTON_TEXT } from '../../const/en';
import { Alert } from 'reactstrap';

interface Props {
  posts: Post[];
  edit: (id: string) => void;
  delete: (id: string) => void;
  error?: any;
}
/**
 * @param data of type Post[]
 * @returns <PostListItem key={post.id}
        post={post}
        edit={props.edit}
        deletePost={props.delete}
      />
 */
const getPostList = (data: Post[], props: Props) => {
  if (!postMessage.length) {
    return null;
  }
  const list = data.map((post) => {
    return (
      <PostListItem
        key={post.id}
        post={post}
        edit={props.edit}
        deletePost={props.delete}
      />
    );
  });
  return list;
};
/**
 * PostBoard displays {@linkcode PostListItem} PostListItem Component
 * Receives edit and delete  as props
 */
const PostList: React.FC<Props> = (props: Props) => {
  const history = useHistory();
  /**
   * Navigates to Add Post page.
   */
  const addPost = () => {
    history.push('/posts/new');
  };
  const postList = props.error ? (
    <Alert color="danger App__Alert">{props.error.message}</Alert>
  ) : (
    getPostList(props.posts, props)
  );
  let classes = ['PostList__Item'];
  classes = props.error ? ['PostList__Item--error', ...classes] : classes;
  return (
    <div data-testid="postlist-container" className="PostList">
      <Row>
        <Col md="10">
          <h1 className="PostList__Heading">{POST_LIST_HEADING}</h1>
        </Col>
      </Row>
      <Row>
        <Col md="6">
          <button
            type="button"
            className="btn--primary"
            title="Add new post"
            onClick={addPost}
          >
            {ADD_POST_BUTTON_TEXT}
          </button>
        </Col>
      </Row>
      <Row>
        <Col className={classes.join(' ')}>{postList}</Col>
      </Row>
    </div>
  );
};
export default PostList;
