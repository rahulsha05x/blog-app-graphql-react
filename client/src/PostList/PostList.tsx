import React from "react";
import './PostList.css';
import PostListItem from "./PostListItem";
import { Row, Col } from "reactstrap";
import { Link } from "react-router-dom";
import { Post } from './Post';

interface Props {
    posts:Post[]
    edit:(id:string)=>void
    delete:(id:string)=>void
}
const getPostList = (data: Post[], props: Props) => {
  if (!postMessage.length) {
    return null;
  }
  const list = data.map((post) => {
    return (
      <PostListItem
        key={post.id}
        post={post}
        edit={props.edit}
        deletePost={props.delete}
      />
    );
  });
  return list;
};
const PostList:React.FC<Props> = (props: Props) => {
  return (
    <div  data-testid='postlist-container' className='PostList'>
      <Row>
        <Col md="10">
          <h1 className="PostList__Heading">Posts</h1>
        </Col>
      </Row>
      <Row>
          <Col md="6">
            <button type='button' className='btn--primary' title='Add new post'>

                <Link to={'/posts/new'}>

                    Add Post
                </Link>
            </button>
          </Col>
      </Row>
      <Row>
        <Col className='PostList__Item'>{getPostList(props.posts, props)}</Col>
      </Row>
    </div>
  );
};
export default PostList;
