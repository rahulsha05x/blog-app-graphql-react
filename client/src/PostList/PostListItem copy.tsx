import React from "react";
import './PostListItem.css';
import { Link, useHistory } from "react-router-dom"
import { Post } from "./Post";
interface Props {
    key:string,
    post:Post,
    edit: (id:string) => void,
    deletePost: (id:string) => void,
}
const PostListItem:React.FC<Props> = ({post,deletePost,edit}: Props) => {
  const { id, title, description } = post;
  const history = useHistory()
  const trimDescription = description.length > 100? `${description.substring(0,150)}...`:description;
  const openPost = (id:any) => {
    history.push(`/posts/${id}`)
  }
  return (
    <article className='PostItem'>
        <section className='PostItem__Card'>
            <h2 className='PostItem__Card--Heading' title={title}>
                <button type='button' data-testid='postitem-heading' className='PostItem__Card--heading' onClick={()=>openPost(id)}>{title}</button>
                {/* <Link to={`/posts/${id}`}>{title}</Link> */}
            </h2>
            <p className='PostItem__Card--Description'>
                {trimDescription}
            </p>
            <article className='PostItem__Card__Button'>
                <button type='button' 
                    className='PostItem__Card__Button--edit btn--primary'
                    onClick={()=>edit(id)}> Edit</button>
                <button 
                    className='PostItem__Card__Button--delete btn--primary' 
                    type='button'
                    onClick={()=>deletePost(id)}>Delete</button>
            </article>
        </section>

    </article>
  );
};

export default PostListItem;
