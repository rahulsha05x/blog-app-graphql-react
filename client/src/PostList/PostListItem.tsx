import React from "react";
import './PostListItem.css';
import { Link } from "react-router-dom"
import { Post } from "./Post";
interface Props {
    post:Post,
    edit: (id:string) => void,
    deletePost: (id:string) => void,
}
const PostListItem:React.FC<Props> = ({post,deletePost,edit}: Props) => {
  const { id, title, description } = post;
  const trimDescription = description.length > 100? `${description.substring(0,150)}...`:description;
  return (
    <article className='PostItem'>
        <section className='PostItem__Card'>
            <h2 className='PostItem__Card--Heading' title={title}>
                <Link to={`/posts/${id}`}>{title}</Link>
            </h2>
            <p className='PostItem__Card--Description'>
                {trimDescription}
            </p>
            <section className='PostItem__Card__Button'>
                <button type='button' 
                    className='PostItem__Card__Button--edit btn--primary'
                    onClick={()=>edit(id)}> Edit</button>
                <button 
                    className='PostItem__Card__Button--delete btn--primary' 
                    type='button'
                    onClick={()=>deletePost(id)}>Delete</button>
            </section>
        </section>

    </article>
  );
};

export default PostListItem;
