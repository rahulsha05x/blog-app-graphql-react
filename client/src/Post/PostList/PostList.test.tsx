import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import PostListItem from "./PostListItem";
import { Route, Router } from "react-router-dom";
import PostList from "./PostList";
import renderer from 'react-test-renderer';

const posts= [
    {
      id: "s-on9Sj",
      title: "New title 121",
      description: "New description 1",
    },
    {
      id: "WzSLQSz",
      title: "New title 12",
      description: "New description 21",
    },
  ];
test("Renders post list with 2 items", async () => {
  render(
        <PostList  posts={posts} edit={jest.fn()} delete={jest.fn()} />
  );
  //screen.debug();
  const items = await screen.findAllByTestId('PostList-Item');
  expect(items).toHaveLength(2);
  
  
});
test("Post item and description text", async () => {
    const component = renderer.create(
          <PostList  posts={posts} edit={jest.fn()} delete={jest.fn()} />
    );
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot();
    const edit = jest.fn();
  });
