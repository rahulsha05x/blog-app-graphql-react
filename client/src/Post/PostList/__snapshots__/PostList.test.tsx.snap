// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post item and description text 1`] = `
<div
  className="PostList"
  data-testid="postlist-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-10"
    >
      <h1
        className="PostList__Heading"
      >
        Posts
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <button
        className="btn--primary"
        onClick={[Function]}
        title="Add new post"
        type="button"
      >
        Add Post
      </button>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="PostList__Item col"
    >
      <article
        className="PostItem"
        data-testid="PostList-Item"
      >
        <section
          className="PostItem__Card"
        >
          <h2
            className="PostItem__Card--Heading"
            title="New title 121"
          >
            <button
              className="PostItem__Card--heading"
              data-testid="postitem-heading"
              onClick={[Function]}
              type="button"
            >
              New title 121
            </button>
          </h2>
          <p
            className="PostItem__Card--Description"
          >
            New description 1
          </p>
          <article
            className="PostItem__Card__Button"
          >
            <button
              className="PostItem__Card__Button--edit btn--primary"
              onClick={[Function]}
              type="button"
            >
               Edit
            </button>
            <button
              className="PostItem__Card__Button--delete btn--primary"
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </article>
        </section>
      </article>
      <article
        className="PostItem"
        data-testid="PostList-Item"
      >
        <section
          className="PostItem__Card"
        >
          <h2
            className="PostItem__Card--Heading"
            title="New title 12"
          >
            <button
              className="PostItem__Card--heading"
              data-testid="postitem-heading"
              onClick={[Function]}
              type="button"
            >
              New title 12
            </button>
          </h2>
          <p
            className="PostItem__Card--Description"
          >
            New description 21
          </p>
          <article
            className="PostItem__Card__Button"
          >
            <button
              className="PostItem__Card__Button--edit btn--primary"
              onClick={[Function]}
              type="button"
            >
               Edit
            </button>
            <button
              className="PostItem__Card__Button--delete btn--primary"
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </article>
        </section>
      </article>
    </div>
  </div>
</div>
`;
